VMware: new module vmware_guest_register_operation

sky-joker opened the pull request:
##### SUMMARY
This module can do the following.

* Register VM to inventory
* Unregister VM from inventory

This is useful when you want to unregister a VM from inventory and register it again.

##### ISSUE TYPE
- New Module Pull Request

sky-joker committed:
Added vmware module vmware_guest_register_operation

sky-joker committed:
fix validate module error code E312

sky-joker committed:
Fix Author account name

sky-joker committed:
Modify path option

sky-joker committed:
Add integration tests

sky-joker committed:
Rename aliases file

sky-joker committed:
fix vcenter/3.6/1 test error

sky-joker commented:
ready_for_review

sky-joker committed:
Delete `msg` key

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+short_description: VM inventory registration operation
+author:
+  - sky-joker (@sky-joker)
+version_added: '2.8'
```
```suggestion
version_added: '2.9'
```

sky-joker replied:
Changed version from 2.8 to 2.9.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+options:
+  datacenter:
+    description:
+    - Destination datacenter for the deploy operation.
```
```suggestion
    - Destination datacenter for the register/unregister operation.
```

sky-joker replied:
Fixed the pointed out line.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+    - '   folder: /folder1/datacenter1/vm'
+    - '   folder: folder1/datacenter1/vm'
+    - '   folder: /folder1/datacenter1/vm/folder2'
+    default: /vm
```
Remove this line.

sky-joker replied:
I deleted the pointed out line.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+            self.module.fail_json(msg="No folder %s matched in the search path : %s" % (self.folder, fullpath),
+                                  details=details)
+
+        if(self.state == "present"):
```
```suggestion
        if self.state == "present":
```

sky-joker replied:
Fixed the pointed out line.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+            fullpath = "%s%s" % (dcpath, self.folder)
+
+        folder_obj = self.content.searchIndex.FindByInventoryPath(inventoryPath="%s" % fullpath)
+        if(not(folder_obj)):
```
```suggestion
        if not folder_obj:
```

sky-joker replied:
Fixed the pointed out line.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+            self.module.fail_json(msg="Cannot find the specified Datacenter: %s" % self.datacenter)
+
+        dcpath = compile_folder_path_for_object(datacenter)
+        if(not(dcpath.endswith("/"))):
```
```suggestion
        if not dcpath.endswith("/"):
```

sky-joker replied:
Fixed the pointed out line.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+        result = dict(changed=False)
+
+        datacenter = self.find_datacenter_by_name(self.datacenter)
+        if(not(datacenter)):
```
```suggestion
        if not datacenter:
```

sky-joker replied:
Fixed the pointed out line.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+                                  details=details)
+
+        if(self.state == "present"):
+            if(find_vm_by_name(self.content, self.name, folder=folder_obj)):
```
Remove parans for every if 
```suggestion
            if find_vm_by_name(self.content, self.name, folder=folder_obj):
```

sky-joker replied:
Modified to use [get_vm](https://github.com/ansible/ansible/blob/4742897635b6e7608f29ca9c49ed894abc51daa4/lib/ansible/module_utils/vmware.py#L880) from [find_vm_by_name](https://github.com/ansible/ansible/blob/4742897635b6e7608f29ca9c49ed894abc51daa4/lib/ansible/module_utils/vmware.py#L228).

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+                                                  pool=resource_pool_obj, host=host_obj)
+
+                try:
+                    wait_for_task(task)
```
```suggestion
                    changed, info = wait_for_task(task)
```

sky-joker replied:
Fixed the pointed out line.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+
+                task = folder_obj.RegisterVM_Task(path=self.path, name=self.name, asTemplate=self.template,
+                                                  pool=resource_pool_obj, host=host_obj)
+
```
```suggestion
changed = False
```

sky-joker replied:
Fixed the pointed out line.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+                except Exception as task_e:
+                    self.module.fail_json(msg=to_native(task_e))
+
+                result.update(changed=True)
```
```suggestion
                result.update(changed=changed)
```

sky-joker replied:
Fixed the pointed out line.

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+                self.module.exit_json(**result)
+
+        else:
+            vm_obj = find_vm_by_name(self.content, self.name, folder=folder_obj)
```
Use `get_vm` API

sky-joker replied:
Modified to use [get_vm](https://github.com/ansible/ansible/blob/4742897635b6e7608f29ca9c49ed894abc51daa4/lib/ansible/module_utils/vmware.py#L880) from [find_vm_by_name](https://github.com/ansible/ansible/blob/4742897635b6e7608f29ca9c49ed894abc51daa4/lib/ansible/module_utils/vmware.py#L228).

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,263 @@
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(datacenter=dict(type="str", default="ha-datacenter"),
+                         cluster_name=dict(type="str"),
+                         folder=dict(type="str", default="/vm"),
```
```suggestion
                         folder=dict(type="str"),
```

sky-joker replied:
Fixed the pointed out line.

Akasurde commented on test/integration/targets/vmware_guest_register_operation/tasks/main.yml:
```diff
@@ -0,0 +1,153 @@
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+- name: Wait for Flask controller to come up online
+  wait_for:
```
Please check [`vmware_guest_find`](https://github.com/ansible/ansible/blob/devel/test/integration/targets/vmware_guest_find/tasks/main.yml) for latest structure of VMware test case.

sky-joker replied:
Modified the test case referring to [vmware_guest_find](https://github.com/ansible/ansible/blob/devel/test/integration/targets/vmware_guest_find/tasks/main.yml).

sky-joker commented:
@Akasurde 

Thank you for the review and advice!
I will fix the advised code.

sky-joker committed:
[update] Made multiple fixes to vmware_guest_register_operation and integration test file

Fixed the following adviced.
- https://github.com/ansible/ansible/pull/54647#discussion_r284596821
- https://github.com/ansible/ansible/pull/54647#discussion_r284597052
- https://github.com/ansible/ansible/pull/54647#discussion_r284597198
- https://github.com/ansible/ansible/pull/54647#discussion_r284597669
- https://github.com/ansible/ansible/pull/54647#discussion_r284597747
- https://github.com/ansible/ansible/pull/54647#discussion_r284597861
- https://github.com/ansible/ansible/pull/54647#discussion_r284598032
- https://github.com/ansible/ansible/pull/54647#discussion_r284598345
- https://github.com/ansible/ansible/pull/54647#discussion_r284598606
- https://github.com/ansible/ansible/pull/54647#discussion_r284598712
- https://github.com/ansible/ansible/pull/54647#discussion_r284598800
- https://github.com/ansible/ansible/pull/54647#discussion_r284599048
- https://github.com/ansible/ansible/pull/54647#discussion_r284599153
- https://github.com/ansible/ansible/pull/54647#discussion_r284599742

Fixed to not use `()` in `if`.
Added option because uuid was required in get_vm api.

jillr commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,266 @@
+                         path=dict(type="str"),
+                         template=dict(type="bool", default=False),
+                         resource_pool=dict(type="str"),
+                         state=dict(type="str", default="present", cohices=["present", "absent"]))
```
```suggestion
                         state=dict(type="str", default="present", choices=["present", "absent"]))
```

sky-joker replied:
Thank you for pointing it out.
Fixed typo.

jillr commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,266 @@
+                result.update(changed=changed)
+                self.module.exit_json(**result)
+
+        else:
```
if self.state == "absent" should be explicitly checked before unregistering the VM, just to be safe.

sky-joker replied:
Thank you for your advice.
I took that advice and added `absent` check.

sky-joker committed:
[update] Fix typo and add absent check.

Fixed the following adviced.
- https://github.com/ansible/ansible/pull/54647#discussion_r285697386
- https://github.com/ansible/ansible/pull/54647#discussion_r285700534

chsnell commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,267 @@
+    - Destination datacenter for the register/unregister operation.
+    - This parameter is case sensitive.
+    default: ha-datacenter
+  cluster_name:
```
These are random comments from someone extremely wanting to see this module (so much that I've pulled it and am testing it out).  Other vmware modules use cluster instead of cluster_name.  Would be nice if it were consistent throughout all the modules.

sky-joker replied:
Thank you for comment!
Modify cluster_name to cluster.

sky-joker replied:
Modified cluster_name parameter name to cluster. :)

chsnell commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,267 @@
+    - 'Examples:'
+    - '    [datastore1] vm/vm.vmx'
+    - '    [datastore1] vm/vm.vmtx'
+  resource_pool:
```
If you set both cluster_name and resource_pool, resource_pool seems to be ignored and the guest is created at the top level of the cluster.  If I leave cluster_name off, and just set resource_pool, the guest creates in the right place.

sky-joker replied:
This assumes that only one of the cluster or resource_pool is specified.
If the `and` conditions for the cluster and resource_pool are required, modify the code.

sky-joker committed:
[update] Modify cluster_name parameter name to cluster and add `type` to DOCUMENTATION

sky-joker committed:
[update] vmware_guest_register_operation test case

sky-joker committed:
change version_added 2.9 to 2.10

goneri commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,277 @@
+description:
+  - This module can register or unregister VMs to the inventory.
+requirements:
+  - python >= 2.6
```
pyvmomi only support 2.7+, so you can safely bump to Python version to 2.7

goneri commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,277 @@
+    description:
+    - Destination datacenter for the register/unregister operation.
+    - This parameter is case sensitive.
+    default: ha-datacenter
```
We don't use any default value for the `datacenter`, See: https://docs.ansible.com/ansible/devel/dev_guide/platforms/vmware_guidelines.html#datacenter-argument-with-esxi

goneri commented on test/integration/targets/vmware_guest_register_operation/tasks/main.yml:
```diff
@@ -0,0 +1,146 @@
+    setup_attach_host: true
+    setup_datastore: true
+    setup_cluster: true
+    setup_virtualmachines: true
```
add `setup_resource_pool: true`, this will create a resource pool called `DC0_C0_RP1`.

goneri commented on test/integration/targets/vmware_guest_register_operation/tasks/main.yml:
```diff
@@ -0,0 +1,146 @@
+  register: resource_pools
+
+- name: get a resource pool
+  set_fact: resource_pool="{{ resource_pools['json'][2].split('/').7 }}"
```
And you can remove the two steps above.

goneri commented on test/integration/targets/vmware_guest_register_operation/tasks/main.yml:
```diff
@@ -0,0 +1,146 @@
+  set_fact: resource_pool="{{ resource_pools['json'][2].split('/').7 }}"
+
+- name: get a vm name
+  set_fact: vm_name="DC0_H0_VM0"
```
Two options here:

- You use `setup_virtualmachines: true` to let `prepare_vmware_tests` prepare the VM for you, and you can use "{{ virtual_machines[0].name }}" or "{{ virtual_machines_in_cluster[0].name }}"
- OR, you remove `setup_virtualmachines: true` the you prepare the VM yourself (please just call it test_vm1 https://docs.ansible.com/ansible/devel/dev_guide/platforms/vmware_guidelines.html#vm-names-should-be-predictable)

goneri reviewed (changes requested):
Nice patch! Thanks @sky-joker.

goneri commented on lib/ansible/modules/cloud/vmware/vmware_guest_register_operation.py:
```diff
@@ -0,0 +1,277 @@
+            else:
+                if self.esxi_hostname:
+                    host_obj = self.find_hostsystem_by_name(self.esxi_hostname)
+                    if not host_obj:
```
This is the 4th `if` level, it probably means you can re-factorize your code a bit, e.g: with a function, of smaller if/else block.

This being said, the code is short enough to be comprehensible, I don't mind merging it like it is today.

sky-joker commented:
Thank you @goneri for pointing it out!
I will cope with the points pointed out.

sky-joker committed:
Made multiple fixes to vmware_guest_register_operation module and test case

Fixe the following.

* https://github.com/ansible/ansible/pull/54647#discussion_r329773449
* https://github.com/ansible/ansible/pull/54647#discussion_r329774024
* https://github.com/ansible/ansible/pull/54647#discussion_r329777135
* https://github.com/ansible/ansible/pull/54647#discussion_r329777432
* https://github.com/ansible/ansible/pull/54647#discussion_r329779517
* https://github.com/ansible/ansible/pull/54647#discussion_r329780945

goneri commented:
The tests pass just fine now, thanks @sky-joker for the quick fix.

sky-joker commented:
Thank you @goneri for merging!

Dinu21 commented:
Can I just use this vmware_guest_register_operation.py as a module in by copying this script in 

/usr/lib/python2.7/site-packages/ansible/modules

Akasurde commented:
@Dinu21 You can - 
* copy required module to `/usr/lib/python2.7/site-packages/ansible/modules/cloud/vmware/` 
or
* create a directory `library` in current working directory and copy-paste required module inside this directory

Please refer - https://docs.ansible.com/ansible/latest/dev_guide/developing_locally.html#adding-a-module-locally for more details.

Dinu21 commented:
> @Dinu21 You can -
> 
> * copy required module to `/usr/lib/python2.7/site-packages/ansible/modules/cloud/vmware/`
>   or
> * create a directory `library` in current working directory and copy-paste required module inside this directory
> 
> Please refer - https://docs.ansible.com/ansible/latest/dev_guide/developing_locally.html#adding-a-module-locally for more details.

@Akasurde Thanks this worked
